scenes/cslib_menu.txt

```diff
@@ -149,121 +167,202 @@
 *temp option_slot_id_9 0
 *temp option_slot_id_10 0
 
-*temp ignore 0
-*set ignore ((cur_page - 1) * items_per_page)
-*temp item_count 0
-*temp cur_option 1
-*temp cur_item "@{invert_order ${{prefix&\"_len\"}}|1}"
-*label _filter_loop
-*if (filter = "")
-	*goto _do_not_filter
-*if ({((prefix&"_")&(filter&"_"))&cur_item} = filter_value)
-	*if (filter_type = "NOT_MATCH")
-		*goto _acc
-*elseif (filter_type != "NOT_MATCH")
-	*goto _acc
-*label _do_not_filter
-*set item_count + 1
-*if (ignore < 1) and (cur_option <= items_per_page)
-	*set option_slot_id[cur_option] cur_item
-	*set cur_option + 1
-*if (ignore > 0)
-	*set ignore -1
-*label _acc
-*if (invert_order)
-	*if (cur_item > 1)
-		*set cur_item - 1
-		*goto _filter_loop
-*else
-	*if (cur_item < {prefix&"_count"})
-		*set cur_item + 1
-		*goto _filter_loop
-*if (item_count = 0) and not(allow_cancel)
-	*bug cslib_menu: menu construction failed as there were 0 matching items — consider allowing cancellation
+*temp total_item_count 0
+*temp cur_option_index 1
+*temp cur_item_index "@{const_invert_order ${{prefix&\"_count\"}}|1}"
+*temp items_off_page ((cur_page - 1) * const_items_per_page)
 
-*label _render
-*temp page_count 0
-*if (item_count < items_per_page)
-	*set page_count 1
+*label filter_loop
+*if (filter != "")
+	*gosub _apply_filter {((prefix&"_")&(filter&"_"))&cur_item_index} filter_value
+	*if not(cslib_ret)
+		*gosub _add_item
 *else
-	*set page_count (item_count / items_per_page)
-*comment this *if might be redundant?
-*if (page_count < 1)
-	*set cslib_ret 0
-	*goto _return
-*if ((page_count modulo 1) > 0)
-	*set page_count ((page_count - (page_count modulo 1)) + 1)
-*temp page_modifier (items_per_page * (cur_page - 1))
+	*gosub _add_item
+*if (items_off_page > 0)
+	*set items_off_page -1
+*gosub _next_item
+*if (cslib_ret)
+	*goto filter_loop
 
+*gosub _set_page_count
+*gosub _populate_page
+*return
+
+*label _populate_page
+*comment populate the choice items and check the guards
+*temp page_item_count 0
+*temp slot 1
+*gosub _init_option_vars
+*label _populate_page_loop
+*if (slot <= const_items_per_page)
+	*if (prefix = "param")
+		*comment special handling for simple menus
+		*temp offset slot+(const_items_per_page * (cur_page -1))
+		*set option_guard[slot] ({prefix&"_count"} >= offset) and (const_items_per_page >= slot)
+		*if (option_guard[slot])
+			*set page_item_count + 1
+			*set option_text[slot] "${param[offset]}"
+			*set option_ret_val[slot] offset
+	*else
+		*set option_guard[slot] ((option_slot_id[slot] > 0) and (const_items_per_page >= slot))
+		*if (option_guard[slot])
+			*set page_item_count + 1
+			*set option_text[slot] "${{(prefix&\"_\")&{\"option_slot_id_\"&slot}}}"
+			*set option_ret_val[slot] option_slot_id[slot]
```

Comment:
this is REALLY hard to follow, CS doesn't help

Reply:
I'm open to suggestions...